### ISSUE TITLE
`gradient` is type-instable

### ISSUE THREAD

**Issue by vchuravy** [FluxML/Zygote.jl#209]
I was looking into a micro-benchmark:

```julia
using Zygote

function matvecother(w, test)
    out = w * test
    loss = sum(abs2.(out .- test))
    # loss = zero(eltype(out))
    # @inbounds for i in eachindex(out)
    #     loss += (test[i] - out[i])^2
    # end
    return loss
end

function testperf()
    weights = rand(Float64,100,100)
    vector = rand(Float64,100)
    loss = 0.0
    for i = 1:50000
        loss = matvecother(weights,vector)
        gradients = gradient(matvecother, weights, vector)
        weights .-= 0.001 .* gradients[1]
        if i%100 == 0
            println(loss)
        end
    end
    return loss
end
```

Which is basically just learning the identity matrix. I was suprised to see that the call to `gradient` results in `gradients::Tuple{Any,Any}` (current Zygote and IRTools master).
Performance get's quite bad if I am using the manually written version of the loss. Am I doing something obviously wrong?

**Comment by MikeInnes** [comment]
At a guess, this is just due to our generic fallback for broadcast being slow. You can fix that by adding a rule [here](https://github.com/FluxML/Zygote.jl/blob/ce8111da692f776f7a8973cf32497319252a0559/src/lib/broadcast.jl#L65). Eventually we should pull in diffrules for this, of course.

**Comment by vchuravy** [comment]
I had a similar issue without broadcast (the commented-out section)

**Comment by CarloLucibello** [comment]
`gradient` is now type-stable here (`@inferred` passes, concrete `Tuple`). Closing as fixed.

_(Triaged against Zygote v0.7.10 on Julia 1.12.6.)_
